{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jfleischer/anaconda/envs/py27/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2818: DtypeWarning: Columns (2,3,5,6,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  if self.run_code(code, result):\n",
      "/Users/jfleischer/anaconda/envs/py27/lib/python2.7/site-packages/ipykernel_launcher.py:9: FutureWarning: 'asobject' is deprecated. Use 'astype(object)' instead\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "%run age_predictors.py\n",
    "\n",
    "def load_data(metadata,datafile):\n",
    "\n",
    "    meta = pd.read_csv(metadata)\n",
    "    \n",
    "    dfo = pd.read_csv(datafile, header=None)\n",
    "    df = dfo.iloc[:,8:].T\n",
    "    pdat = df.iloc[:,0].asobject\n",
    "\n",
    "    uid = []\n",
    "    ageType = []\n",
    "    age = []\n",
    "    source = []\n",
    "    condition = []\n",
    "    ethnicity = []\n",
    "    gender = []\n",
    "    datep = []\n",
    "    whop = []\n",
    "    \n",
    "    #print dfo.shape\n",
    "    \n",
    "    for el in pdat:\n",
    "        tid = el.split('_')[0]\n",
    "        mdata = meta[meta['Tracking ID']==tid]\n",
    "        uid.append(el)  \n",
    "    \n",
    "        x = mdata['Age'].iloc[0].lower()\n",
    "        if 'fw' in x:\n",
    "            ageType.append('fw') \n",
    "            m = re.search('\\d{1,2} ?fw',x)\n",
    "            age.append(int(m.group(0)[:-2]))\n",
    "        elif 'nb' in x:\n",
    "            ageType.append('nb')\n",
    "            age.append(0)\n",
    "        elif 'da' in x:\n",
    "            ageType.append('nb')\n",
    "            age.append(0)\n",
    "        elif 'yr' in x:\n",
    "            ageType.append('yr')\n",
    "            m = re.search('\\d{1,3} ?yr',x)\n",
    "            # maybe use months sometime too, but not now\n",
    "            # m2 = re.search('\\d{1,2} ?mo',x) \n",
    "            age.append(int(m.group(0)[:-2]))\n",
    "        else:\n",
    "            print(\"OOOPS! \" + x)\n",
    "                \n",
    "\n",
    "        x = mdata['Condition'].iloc[0].lower()      \n",
    "#        could be any of these types:\n",
    "#                lab derived fibroblasts\n",
    "#                normal\n",
    "#                direct skin biopsy\n",
    "#                control for hgps\n",
    "#                relative of hgps\n",
    "#                control for ism\n",
    "#                ism\n",
    "#                ism(sorted)\n",
    "#                ism(hgps)\n",
    "#                ism(relative of hgps)\n",
    "#                heated direct skin biopsy\n",
    "        condition.append(x)\n",
    "        if 'fibroblast' in x:\n",
    "            source.append('fibroblast')\n",
    "        elif 'biopsy' in x:\n",
    "            source.append('biopsy')\n",
    "        else:\n",
    "            source.append('coriell')\n",
    "            \n",
    "        gender.append( mdata['Gender'].iloc[0].lower() )\n",
    "        ethnicity.append( mdata['Race/Ethnicity'].iloc[0].lower() )\n",
    "        oname = mdata['Original Name'].iloc[0]\n",
    "        datep.append( '20'+oname[:2]+'-'+oname[2:4]+'-'+oname[4:6] )\n",
    "        whop.append( oname.split('/')[1].split('_')[-1] )\n",
    "   \n",
    "            \n",
    "            \n",
    "    #\n",
    "    #df = df.drop(0)\n",
    "    #del df['patient data']\n",
    "    #print len(ageType),len(age), len(condition), len(source), len(df)\n",
    "    \n",
    "    \n",
    "    df.index = uid\n",
    "    df.drop(df.columns[0],axis=1,inplace=True)\n",
    "    df.columns = dfo.iloc[1:,0]\n",
    "    trnum=[]\n",
    "    for el in df.index:\n",
    "        xx = el.split('_')[0]\n",
    "        if xx=='pdl33': # get rid of the non-numeric tracking number which indicates #147 cells cultured to population doubling level 33\n",
    "            trnum.append(33147)\n",
    "        else:\n",
    "            trnum.append(int(xx))\n",
    "    df['tracking number'] = trnum\n",
    "    df['ethnicity'] = ethnicity\n",
    "    df['gender'] = gender\n",
    "    df['source'] = source\n",
    "    df['condition'] = condition\n",
    "    df['age type'] = ageType\n",
    "    df['age'] = age\n",
    "    df['date processed'] = datep\n",
    "    df['who processed'] = whop\n",
    "    \n",
    "    # rearrange so its easy to make a data matrix\n",
    "    cols = df.columns.tolist()\n",
    "    ncols = cols[-9:] + cols[:-9]\n",
    "    df = df[ncols]   \n",
    "\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "alldata = load_data('MetaDataSheet.csv','fpkm.csv')\n",
    "\n",
    "\n",
    "normals = (alldata.condition == 'normal') \n",
    "progeria = (alldata.condition == 'hgps')  \n",
    "genes = alldata.iloc[:,9:].astype(float) # get only the gene expression levels, exclude the metadata\n",
    "ages = alldata[normals].age \n",
    "age_prog = alldata[progeria].age\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "alld2 = pd.read_csv('../raw data/fpkm-Sept1-removeISM.csv',header=None)\n",
    "#parse the gene symbol out of the full columns from raw HOMER output\n",
    "gene_symbol = alld2.iloc[1:,7].apply(lambda x: x.split('|')[0]) \n",
    "transcripts = alld2.iloc[1:,0]\n",
    "\n",
    "# make sure the raw data transcripts matches the transcripts from the usual fpkm table\n",
    "print(gene_symbol.shape, genes.columns.shape)\n",
    "print((transcripts == genes.columns).all())\n",
    "\n",
    "all_transcript_labelled_data = alldata.reset_index().set_index(['tracking number','age','condition','ethnicity','gender','source','date processed','who processed']).drop(['index','age type'],axis=1)\n",
    "all_gene_labelled_data = all_transcript_labelled_data.copy()\n",
    "all_gene_labelled_data.columns = gene_symbol # replace transcript labels with gene labels\n",
    "all_gene_labelled_data.to_csv('all_gene_labelled_data.csv')\n",
    "all_transcript_labelled_data.to_csv('all_transcript_labelled_data.csv')\n",
    "\n",
    "\n",
    "transcript_labelled_data = alldata[normals].reset_index().set_index(['tracking number','age','condition']).iloc[:,5:]\n",
    "gene_labelled_data = transcript_labelled_data.copy()\n",
    "gene_labelled_data.columns = gene_symbol\n",
    "gene_labelled_data.to_csv('gene_labelled_data.csv')\n",
    "transcript_labelled_data.to_csv('transcript_labelled_data.csv')\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
